crates/holochain_integrity_types/src/capability/grant.rs

```diff
@@ -111,13 +111,19 @@ impl CapGrant {
                     // The grant has no assignees so is always valid…
                     _ => true,
                 }
-                // The secret needs to match…
-                && match access {
-                    // Unless the extern is unrestricted.
-                    CapAccess::Unrestricted => true,
-                    // note the PartialEq implementation is constant time for secrets
-                    CapAccess::Transferable { secret, .. } => check_secret.map(|given| secret == given).unwrap_or(false),
-                    CapAccess::Assigned { secret, .. } => check_secret.map(|given| secret == given).unwrap_or(false),
```

Comment:
Well, if this function named `is_valid` is being called for a `Transferable` cap grant but with `check_secret == None` then the result cannot really be represented with a boolean in the first place, can it? It's rather an "undetermined" state - so maybe the function should be called `is_validated` instead (small joke ;-) )?
So I see how one could reason to return false if no secret is provided for a Transferable cap grant to protect against this function accidentally being called for a transferable grant without providing the secret.

Reply:
@maackle after reading your comment in light of @matthme 's comment I think I see what you're saying. The effect of skipping the check would be that the loop that checks all the grants on the source chain would also come up with nothing and return false, right? I'm wondering when are the times that the secret would be `None` -- AFAICT it's only when the caller didn't supply one, in which case I think this check should probably return `false` rather than getting skipped (although the effect is the same -- access denied).

Reply:
> If check_secret is None, then this function would always just return false

For unrestricted cap grants it returns `true`, and the other two variants required a secret, so you couldn't skip the secret check. It seemed correct to me.

If the secret becomes optional for assigned cap grants, the logic can no longer be used. But the update is wrong, because a transferable cap grant still requires a secret, and now if none is provided, the secret check is skipped.

Reply:
> this check should probably return false rather than getting skipped (although the effect is the same -- access denied).

In order to properly implement 'skipped' logic in a way that doesn't let [`SourceChain::valid_cap_grant`](https://github.com/holochain/holochain/blob/c745c93471d697cbbe2f1f3478a0a8dacbf00ffc/crates/holochain_state/src/source_chain.rs#L652-L657) allow unauthorised people access a function by supplying a `None` secret, this function would have to have a return signature of `Option<bool>` and let its consumer treat `None` as equivalent to `false` in the loop... (which I guess is what you're saying @matthme ) I'd just as soon return `false` if a required secret wasn't supplied.

Reply:
I made the mistake of interpreting `check_secret` as an imperative i.e. if it's Some, we want to check the secret, if it's None, we don't want to check the secret, and it's the responsibility of the caller to decide whether they want to check the secret or not. I did this too casually, looking more deeply I see that check_secret comes directly from the zome caller, so we can't give them that responsibility.

Let me try again.

Reply:
LGTM. It would be great to have tests for this method. Are you willing to add them? Otherwise I would.

Reply:
Test written.